### PR TITLE
Train classifier

### PR DESCRIPTION
**Pull request by antoinemrcr**
Add a script to extract features and train a classifier on top of these features.
Update the gesture recognition script to use a trained classifier.

### CODE REVIEW

File: realtimenet/finetune_utils.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+                    video_fps = video_source.get_sample_rate()
+
+                    while True:
+                        images = video_source.get_image()
+                        if images is None:
+                            break
+                        else:
+                            image, image_rescaled = images
+                            frames.append(image_rescaled)
+                    frames = uniform_frame_sample(np.array(frames), 16/video_fps)
```
Comment (guillaumebrg) [inline realtimenet/finetune_utils.py:97]: 16 --> inference_engine.fps

File: scripts/train_classifier.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    features_dir = os.path.join(path_in, "features_train")
+    classes = os.listdir(videos_dir)
+    classes = [x for x in classes if not x.startswith('.')]
+
+
+    # finetune the model
+    extract_features(path_in, classes, net, num_layer_finetune, use_gpu)
+
+    y_train, y_valid = [], []
+    X_train, X_valid = [], []
```
Comment (guillaumebrg) [inline scripts/train_classifier.py:100]: Doesn't seem to be used.

File: scripts/gesture_recognition.py [hunk 3]
```diff
@@ -33,20 +37,35 @@
...
     use_gpu = args['--use_gpu']
 
     # Load feature extractor
     feature_extractor = feature_extractors.StridedInflatedEfficientNet()
     checkpoint = engine.load_weights('resources/strided_inflated_efficientnet.ckpt')
-    feature_extractor.load_state_dict(checkpoint)
-    feature_extractor.eval()
 
     # Load a logistic regression classifier
+    if custom_classifier is not None:
```
Comment (guillaumebrg) [inline scripts/gesture_recognition.py:49]: Note to self: Should we have a dedicated script for this?

File: scripts/train_classifier.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    feature_extractor.eval()
+
+    # Concatenate feature extractor and met converter
+    net = feature_extractor
+
+    # list the labels from the training directory
+    videos_dir = os.path.join(path_in, "videos_train")
+    features_dir = os.path.join(path_in, "features_train")
+    classes = os.listdir(videos_dir)
+    classes = [x for x in classes if not x.startswith('.')]
```
Comment (guillaumebrg) [inline scripts/train_classifier.py:93]: Note to self: instead of organising videos into sub-folders, would it make more sense to have one folder with all the videos, one annotation file with the label for each video, and take care of splitting the data in the code?

File: realtimenet/camera.py [hunk 5]
```diff
@@ -42,6 +42,9 @@ def pad_to_square(self, img):
         pad_right = square_size - img.shape[1] - pad_left
         return cv2.copyMakeBorder(img, pad_top, pad_bottom, pad_left, pad_right, cv2.BORDER_CONSTANT)
 
+    def get_sample_rate(self):
+        return self._cam.get(cv2.CAP_PROP_FPS)
```
Comment (guillaumebrg) [inline realtimenet/camera.py:46]: Check: does that actually return the framerate of the encoded video?

**Review by guillaumebrg — COMMENTED** [review]
Did a first round of reviewing. I'm going to play a little bit with that script now.

File: realtimenet/finetune_utils.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+import glob
+import numpy as np
+import torch
+import torch.optim as optim
+import torch.nn as nn
+import os
+
+
+class FeaturesDataset(torch.utils.data.Dataset):
+    """Face Landmarks dataset."""
```
[the remark this answers is not in the document]
Comment (antoinemrcr) [reply realtimenet/finetune_utils.py:12]: pytorch documentation

[the remark this answers is not in the document]
Comment (antoinemrcr) [reply realtimenet/finetune_utils.py:12]: but i only really kept just the method do modify

File: realtimenet/finetune_utils.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        self.num_timesteps = num_timesteps
+
+    def __len__(self):
+        return len(self.files)
+
+    def __getitem__(self, idx):
+        feature = np.load(self.files[idx])
+        num_preds = feature.shape[0]
+        if num_preds <= self.num_timesteps:
+            return self.__getitem__(idx + 1)
```
[the remark this answers is not in the document]
Comment (antoinemrcr) [reply realtimenet/finetune_utils.py:26]: i modified the extraction feature to not extract if the video duration is too short.

File: realtimenet/feature_extractors/mobilenet.py [hunk 8]
```diff
@@ -30,9 +31,10 @@ def temporal_footprint(self):
         return effective_kernel_size - self.stride_temporal
 
     def forward(self, x):
-        if self.internal_state is None:
-            self.initialize_internal_state(x)
-        x = self.pad_internal_state(x)
+        if self.internal_padding:
```
[the remark this answers is not in the document]
Comment (antoinemrcr) [reply realtimenet/feature_extractors/mobilenet.py:34]: Tried to use apply to change the value when we create the network

**Review by guillaumebrg — APPROVED** [review]
Should be good to go :+1:
